File: ios/RecipeClipperTests/Fakes/FakeListRepository.swift
import Combine
import Foundation
@testable import RecipeClipper

/// A fake that models membership as real state rather than only recording calls: the
/// behaviour worth testing is a round trip (ticking a list writes, and the publisher the sheet
/// is watching re-emits with the box now checked). `recipeCount` and `containsRecipe` are
/// derived from `membership` exactly as the SQL derives them, so a `recipeCount` staged on a
/// list is ignored — stage membership instead.
final class FakeListRepository: ListRepository {
    struct Membership: Hashable {
        let recipeId: Int64
        let listId: Int64
    }

    /// The lists that exist. `recipeCount` / `containsRecipe` on these are overwritten.
    let lists = CurrentValueSubject<[RecipeList], Never>([])
    let membership = CurrentValueSubject<Set<Membership>, Never>([])
    /// What `observeRecipesIn` emits, regardless of the list asked for.
    let recipesIn = CurrentValueSubject<[RecipeSummary], Never>([])

    private(set) var createCalls: [(name: String, recipeId: Int64?)] = []
    private(set) var renameCalls: [(listId: Int64, name: String)] = []
    private(set) var deleteCalls: [Int64] = []
    private(set) var recipesInQueries: [Int64] = []

    private var nextId: Int64 = 100

    func stageMembership(_ pairs: [(recipeId: Int64, listId: Int64)]) {
        membership.send(Set(pairs.map { Membership(recipeId: $0.recipeId, listId: $0.listId) }))
    }

    func observeLists() -> AnyPublisher<[RecipeList], Never> {
        lists.combineLatest(membership)
            .map { all, rows in all.map { Self.withCounts($0, rows, recipeId: nil) } }
            .eraseToAnyPublisher()
    }

    func observeListsFor(recipeId: Int64) -> AnyPublisher<[RecipeList], Never> {
        lists.combineLatest(membership)
            .map { all, rows in all.map { Self.withCounts($0, rows, recipeId: recipeId) } }
            .eraseToAnyPublisher()
    }

    func observeRecipesIn(listId: Int64) -> AnyPublisher<[RecipeSummary], Never> {
        recipesInQueries.append(listId)
        return recipesIn.eraseToAnyPublisher()
    }

    @MainActor func setMembership(recipeId: Int64, listId: Int64, inList: Bool) async {
        var rows = membership.value
        let row = Membership(recipeId: recipeId, listId: listId)
        if inList { rows.insert(row) } else { rows.remove(row) }
        membership.send(rows)
    }

    @MainActor
    @discardableResult
    func createList(name: String, addRecipeId: Int64?) async -> Int64 {
        createCalls.append((name, addRecipeId))
        let id = nextId
        nextId += 1
        lists.send(lists.value + [RecipeList(id: id, name: name, isBuiltIn: false, isFavorites: false, recipeCount: 0)])
        if let addRecipeId {
            membership.send(membership.value.union([Membership(recipeId: addRecipeId, listId: id)]))
        }
        return id
    }

    @MainActor func rename(listId: Int64, name: String) async {
        renameCalls.append((listId, name))
        lists.send(lists.value.map { list in
            var list = list
            if list.id == listId { list.name = name }
            return list
        })
    }

    @MainActor func deleteList(listId: Int64) async {
        deleteCalls.append(listId)
        // Favorites is refused in SQL, so the fake refuses it here for the same reason.
        if lists.value.contains(where: { $0.id == listId && $0.isFavorites }) { return }
        lists.send(lists.value.filter { $0.id != listId })
        membership.send(membership.value.filter { $0.listId != listId })
    }

    private static func withCounts(_ list: RecipeList, _ rows: Set<Membership>, recipeId: Int64?) -> RecipeList {
        var list = list
        list.recipeCount = rows.filter { $0.listId == list.id }.count
        list.containsRecipe = recipeId.map { rows.contains(Membership(recipeId: $0, listId: list.id)) } ?? false
        return list
    }
}

File: app/src/test/java/com/example/recipeclipper/fake/FakeListRepository.kt
package com.example.recipeclipper.fake

import com.example.recipeclipper.data.ListRepository
import com.example.recipeclipper.data.model.RecipeList
import com.example.recipeclipper.data.model.RecipeSummary
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.map

/**
 * A hand-written fake that actually models membership rather than only recording calls.
 *
 * The behaviour worth testing here is a round trip — ticking a list writes, and the flow the
 * sheet is collecting re-emits with the box now checked — and a fake that just recorded
 * `setMembership` would prove only half of that. So [membership] is real state, and
 * `recipeCount` and `containsRecipe` are derived from it exactly as the SQL derives them.
 *
 * That means the `recipeCount` a test puts in a staged [lists] entry is ignored: stage
 * [membership] instead and the counts follow.
 */
class FakeListRepository : ListRepository {

    /** The lists that exist. `recipeCount`/`containsRecipe` on these are overwritten. */
    val lists = MutableStateFlow<List<RecipeList>>(emptyList())

    /** recipeId to listId, one pair per membership row. */
    val membership = MutableStateFlow<Set<Pair<Long, Long>>>(emptySet())

    /** What [observeRecipesIn] emits, regardless of the list asked for. */
    val recipesIn = MutableStateFlow<List<RecipeSummary>>(emptyList())

    /** Name and the recipe to add (or null), for every [createList] call, in order. */
    val createCalls = mutableListOf<Pair<String, Long?>>()

    /** Every [rename] call, in order. */
    val renameCalls = mutableListOf<Pair<Long, String>>()

    /** Every list id [deleteList] was called with, in order. */
    val deleteCalls = mutableListOf<Long>()

    /** Every list id [observeRecipesIn] was asked for, in order. */
    val recipesInQueries = mutableListOf<Long>()

    private var nextId = 100L

    override fun observeLists(): Flow<List<RecipeList>> =
        combine(lists, membership) { all, rows -> all.map { it.withCounts(rows, recipeId = null) } }

    override fun observeListsFor(recipeId: Long): Flow<List<RecipeList>> =
        combine(lists, membership) { all, rows -> all.map { it.withCounts(rows, recipeId) } }

    override fun observeRecipesIn(listId: Long): Flow<List<RecipeSummary>> {
        recipesInQueries += listId
        return recipesIn.map { it }
    }

    override suspend fun setMembership(recipeId: Long, listId: Long, inList: Boolean) {
        membership.value = if (inList) {
            membership.value + (recipeId to listId)
        } else {
            membership.value - (recipeId to listId)
        }
    }

    override suspend fun createList(name: String, addRecipeId: Long?): Long {
        createCalls += name to addRecipeId
        val id = nextId++
        lists.value = lists.value + RecipeList(
            id = id,
            name = name,
            isBuiltIn = false,
            isFavorites = false,
            recipeCount = 0
        )
        if (addRecipeId != null) membership.value = membership.value + (addRecipeId to id)
        return id
    }

    override suspend fun rename(listId: Long, name: String) {
        renameCalls += listId to name
        lists.value = lists.value.map { if (it.id == listId) it.copy(name = name) else it }
    }

    override suspend fun deleteList(listId: Long) {
        deleteCalls += listId
        // Favorites is refused in SQL, so the fake refuses it here for the same reason.
        if (lists.value.any { it.id == listId && it.isFavorites }) return
        lists.value = lists.value.filterNot { it.id == listId }
        membership.value = membership.value.filterNot { it.second == listId }.toSet()
    }

    private fun RecipeList.withCounts(rows: Set<Pair<Long, Long>>, recipeId: Long?) = copy(
        recipeCount = rows.count { it.second == id },
        containsRecipe = recipeId != null && (recipeId to id) in rows
    )
}
